cosmo_tester/test_suites/snapshots/multi_tenant_snapshot_test.py

```diff
@@ -13,141 +13,182 @@
+    check_deployments,
+    check_from_source_plugin,
+    check_plugins,
+    _cluster,
+    confirm_manager_empty,
+    create_noinstall_helloworld,
+    create_snapshot,
+    delete_manager,
+    download_snapshot,
+    get_deployments_list,
+    get_secrets_list,
+    NOINSTALL_DEPLOYMENT_ID,
+    remove_and_check_deployments,
+    restore_snapshot,
+    set_client_tenant,
+    SNAPSHOT_ID,
+    upgrade_agents,
+    upload_and_install_helloworld,
+    upload_snapshot,
+    upload_test_plugin,
 )
 
-TENANTS = ['tenant_a', 'tenant_b', 'tenant_c']
-
-
-# TODO: Add version as master progresses
-@pytest.fixture(scope='module', params=['master'])
-def cluster(request, cfy, ssh_key, module_tmpdir, attributes, logger):
-    managers = (
-        MANAGERS[request.param](),
-        MANAGERS['master'](upload_plugins=False),
-    )
-
-    cluster = CloudifyCluster.create_image_based(
-            cfy,
-            ssh_key,
-            module_tmpdir,
-            attributes,
-            logger,
-            managers=managers,
-            )
 
+def test_restore_snapshot_and_agents_upgrade_multitenant(
+        cfy, cluster_multitenant, attributes, logger, tmpdir):
+    local_snapshot_path = str(tmpdir / 'snapshot.zip')
+    new_tenants = ('tenant1', 'tenant2')
+    hello_tenants = ('default_tenant', new_tenants[0])
+    noinstall_tenants = new_tenants
+
+    old_manager = cluster_multitenant.managers[0]
+    new_manager = cluster_multitenant.managers[1]
+    hello_vms = cluster_multitenant.managers[2:]
+
+    hello_vm_mappings = {
+        hello_tenants[0]: cluster_multitenant.managers[2],
+        hello_tenants[1]: cluster_multitenant.managers[3],
+    }
+
+    confirm_manager_empty(new_manager)
+
+    make_tenants(old_manager, logger, tenants=new_tenants)
+    tenants = ['default_tenant']
+    tenants.extend(new_tenants)
+
+    for tenant in hello_tenants:
+        upload_and_install_helloworld(attributes, logger, old_manager,
+                                      hello_vm_mappings[tenant],
+                                      tmpdir, tenant=tenant, prefix=tenant)
+    for tenant in noinstall_tenants:
+        create_noinstall_helloworld(old_manager, logger, tenant=tenant)
+
+    create_tenant_secrets(old_manager, tenants, logger)
+
+    old_plugins = {
+        tenant: upload_test_plugin(old_manager, logger, tenant)
+        for tenant in tenants
+    }
+    old_secrets = {
+        tenant: get_secrets_list(old_manager, tenant)
+        for tenant in tenants
+    }
+    old_deployments = {
+        tenant: get_deployments_list(old_manager, tenant)
+        for tenant in tenants
+    }
+
+    create_snapshot(old_manager, SNAPSHOT_ID, attributes, logger)
+    download_snapshot(old_manager, local_snapshot_path, SNAPSHOT_ID, logger)
+    upload_snapshot(new_manager, local_snapshot_path, SNAPSHOT_ID, logger)
+
+    restore_snapshot(new_manager, SNAPSHOT_ID, cfy, logger)
+
+    assert_hello_worlds(hello_vms, installed=True, logger=logger)
```

Comment:
I'm adding a check afterwards as well. It's a fairly quick check (except when it fails, though even then I think we're talking 30 seconds at most).

That'll mean that it'll allow narrowing down the failure to before or after the agent install if there is strangeness.